idseq_pipeline/commands/non_host_alignment_functions.py

```diff
@@ -717,17 +839,81 @@ def run_gsnapl_chunk(part_suffix, remote_home_dir, remote_index_dir, remote_work
             execute_command(scp(key_path, remote_username, gsnapl_instance_ip, remote_outfile, local_outfile))
+            execute_command(scp(key_path, remote_username, gsnapl_instance_ip, multihit_remote_outfile, multihit_local_outfile))
         # Deduplicate m8 input. Sometimes GSNAPL outputs multiple consecutive lines for same original read and same accession id. Count functions expect only 1 (top hit).
         execute_command("echo '' >> %s;" % local_outfile) # add a blank line at the end of the file so S3 copy doesn't fail if output is empty
         deduplicate_m8(os.path.join(CHUNKS_RESULT_DIR, outfile_basename), os.path.join(CHUNKS_RESULT_DIR, dedup_outfile_basename))
         with iostream:
             execute_command("aws s3 cp --quiet %s/%s %s/" % (CHUNKS_RESULT_DIR, dedup_outfile_basename, SAMPLE_S3_OUTPUT_CHUNKS_PATH))
+
+    # Deduplicate multihit m8 by using taxonomy info
+    multihit_summary_file = CHUNKS_RESULT_DIR + "/summary-multihit-" + outfile_basename
+    dedup_multihit_local_outfile = CHUNKS_RESULT_DIR + "/dedup-multihit-" + outfile_basename
+    call_hits_m8(multihit_local_outfile, dedup_multihit_local_outfile, multihit_summary_file)
+    with iostream:
+        execute_command("aws s3 cp --quiet %s %s/" % (dedup_multihit_local_outfile, SAMPLE_S3_OUTPUT_CHUNKS_PATH))
+        execute_command("aws s3 cp --quiet %s %s/" % (multihit_summary_file, SAMPLE_S3_OUTPUT_CHUNKS_PATH))
+
     with iostream:
         execute_command("sed -i '$ {/^$/d;}' %s" % os.path.join(CHUNKS_RESULT_DIR, dedup_outfile_basename)) # remove blank line from end of file
     write_to_log("finished alignment for chunk %s" % chunk_id)
     return os.path.join(CHUNKS_RESULT_DIR, dedup_outfile_basename)
 
 
+def call_hits_m8(input_m8, output_m8, output_summary):
+    # Get lineage info
+    lineage_path = fetch_reference(LINEAGE_SHELF)
+    lineage_map = shelve.open(lineage_path)
+    accession2taxid_path = fetch_reference(ACCESSION2TAXID)
+    accession2taxid_dict = shelve.open(accession2taxid_path)
+    # Helper functions
+    def add_taxid_hits(accession, hits):
+        taxid = accession2taxid_dict.get(accession.split(".")[0], "NA")
+        species_taxid, genus_taxid, family_taxid = lineage_map.get(taxid, ("-100", "-200", "-300"))
+        hits["species"] += [species_taxid]
+        hits["genus"] += [genus_taxid]
+        hits["family"] += [family_taxid]
+        return hits
+    def call_hit_level(read_id, hits):
+        for i, level in enumerate(["species", "genus", "family"]):
+            taxids = hits[level]
+            taxids = [t for t in taxids if int(t) >= 0] # do not consider unclassified hits. TO DO: decide if that's an improvement
+            n = len(set(taxids)) # number of distinct taxids at this level
+            if n == 1:
+                return i+1, taxids[0]
+        return -1, -1
+    # Deduplicate m8 and summarize hits
+    read_ids = execute_command_with_output("grep -v '^#' %s | cut -f1" % input_m8).split("\n")
```

Comment:
Just use `subprocess.check_output` here instead of `execute_command_with_output`.   The latter is meant for heavy weight infrequently called commands.